sycl/include/sycl/device_selector.hpp

```diff
@@ -31,7 +31,8 @@ class filter_selector;
 /// \sa device
 ///
 /// \ingroup sycl_api_dev_sel
-class __SYCL_EXPORT device_selector {
+class __SYCL_EXPORT __SYCL2020_DEPRECATED(
+    "Use Callable instead to select device.") device_selector {
```

Comment:
SYCL2020 doesn't require these compatibility types to be inherited from device selector. I think it would simplify the code by implementing them as 

```
struct /* export/deprecated */ default_selector {
  int operator(const device &dev) {
    return default_selector_v(dev);
  }
};
```

I.e., without inheritance and with inline definition of `operator()` vs. keeping it in .cpp file.

I think the reason of having it in current state is `std::is_invocable` being C++17. But, IMO, it would be better either not to guard the ctor with `enable_if` or provide implementation of `is_invocable` in our compatibility header.

Or is the reason that we want to provide the user ability to create custom selector while still inheriting from the base `device_selector`?

Reply:
We might be able to provide our own C++14 compatible `is_invocable` but that's outside the scope of this PR. 

Here we are just marking the existing SYCL 1.2.1 device_selectors (and those functions that use them) as deprecated, and expunging their usage from the code base. We'll remove the `device_selector` class and its subclasses at a later time.